oprf/group/group.go

```diff
@@ -0,0 +1,206 @@
+		sign := Sgn0(p.y)
+		// select correct tag
+		e := int(Equals(sign, One).Int64())
+		tag = subtle.ConstantTimeSelect(e, 2, 3)
+	}
+
+	return append([]byte{byte(tag)}, b...)
+}
+
+// checkBytes checks that the number of bytes corresponds to the correct
+// curve type and serialization tag that is present
+func isCompressed(in []byte, expLen int) (bool, error) {
+	tag := in[0]
+	com := false
+
+	switch tag {
+	case 2, 3:
+		if expLen < len(in)-1 {
+			return false, errors.New("error deserializing group element")
+		}
+		com = true
+	case 4:
+		if expLen*2 < len(in)-1 {
+			return false, errors.New("error deserializing group element")
+		}
+	default:
+		return false, errors.New("error deserializing group element")
+	}
+
+	return com, nil
+}
+
+// Deserialize an octet-string into a valid Element object.
+func (p *Element) Deserialize(in []byte) error {
+	byteLen := p.c.ByteLength()
+
+	com, err := isCompressed(in, byteLen)
+	if err != nil {
+		return err
+	}
+
+	if !com {
+		p.x = new(big.Int).SetBytes(in[1 : byteLen+1])
+		p.y = new(big.Int).SetBytes(in[byteLen+1:])
+
+		if !p.IsValid() {
+			return errors.New("invalid deserialization")
+		}
+	} else {
+		order := p.c.curve.Params().P
+		var y2 *big.Int
+		x := new(big.Int).SetBytes(in[1:])
+
+		x2 := new(big.Int).Exp(x, Two, order)
+		x2a := new(big.Int).Add(x2, big.NewInt(-3))
+		x3 := new(big.Int).Mul(x2a, x)
+		x3ab := new(big.Int).Add(x3, p.c.curve.Params().B)
+		y2 = new(big.Int).Mod(x3ab, order)
+
+		a1 := new(big.Int).Add(p.c.curve.Params().P, One)
+		a2 := new(big.Int).ModInverse(big.NewInt(4), p.c.curve.Params().P)
+		m := new(big.Int).Mul(a1, a2)
+		sqrtExp := new(big.Int).Mod(m, p.c.curve.Params().P)
+
+		// construct y coordinate with correct sign
+		y := new(big.Int).Exp(y2, sqrtExp, order)
+		parity := Equals(big.NewInt(int64(in[0])), Two)
+		yParity := Equals(Sgn0(y), One)
+		y = cMov(new(big.Int).Mul(y, MinusOne), y, Equals(parity, yParity))
+
+		p.x = new(big.Int).Mod(x, order)
+		p.y = new(big.Int).Mod(y, order)
+
+		if !p.IsValid() {
+			return errors.New("invalid deserialization")
+		}
+
+		p.com = true
+	}
+
+	return nil
+}
+
+// Equal returns a bool indicating whether two Elements are equal.
+func (p *Element) Equal(q *Element) bool {
+	return (p.x.Cmp(q.x) == 0) && (p.y.Cmp(q.y) == 0)
+}
+
+// Scalar is an struct representing a field element
+type Scalar struct {
+	c *Ciphersuite
+	x *big.Int
+}
+
+// NewScalar generates a new scalar.
+func NewScalar(c *Ciphersuite) *Scalar {
+	s := &Scalar{c, new(big.Int)}
+	return s
+}
+
+// Set sets the scalar to a value.
+func (s *Scalar) Set(x []byte) *Scalar {
+	s.x.SetBytes(x)
+	return s
+}
+
+// Inv sets the Scalar to its multiplicative inverse.
+func (s *Scalar) Inv() *Scalar {
+	n := s.c.Order()
+	inv := new(big.Int).ModInverse(s.x, n.x)
+
+	rInv := NewScalar(s.c)
+	rInv.x.Set(inv)
+
+	return rInv
+}
+
+// Serialize the Scalar into a byte slice.
+func (s *Scalar) Serialize() []byte {
+	l := s.c.ByteLength()
+	bytes := s.x.Bytes()
```

Comment:
Why can't we just return `s.x.Bytes()`?

Reply:
I think this is needed for p521, but I have to check.

Reply:
Removed as I couldn't find the reference. If later there is some problems, I'm sure they can be reported.